lib/spack/spack/solver/asp.py

```diff
@@ -1404,43 +1464,51 @@ def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=
         Returns:
             int: id of the condition created by this function
         """
+        # Defensive copy of string or spec
         named_cond = required_spec.copy()
-        named_cond.name = named_cond.name or name
-        assert named_cond.name, "must provide name for anonymous conditions!"
-
+        if isinstance(named_cond, spack.spec.Spec):
+            named_cond.name = named_cond.name or name
+            name = named_cond.name
+        assert name, "must provide name for anonymous conditions!"
         # Check if we can emit the requirements before updating the condition ID counter.
         # In this way, if a condition can't be emitted but the exception is handled in the caller,
         # we won't emit partial facts.
 
         condition_id = next(self._condition_id_counter)
-        self.gen.fact(fn.pkg_fact(named_cond.name, fn.condition(condition_id)))
+        self.gen.fact(fn.pkg_fact(name, fn.condition(condition_id)))
         self.gen.fact(fn.condition_reason(condition_id, msg))
 
-        cache = self._trigger_cache[named_cond.name]
+        cache = self._trigger_cache[name]
 
         named_cond_key = str(named_cond)
         if named_cond_key not in cache:
             trigger_id = next(self._trigger_id_counter)
-            requirements = self.spec_clauses(named_cond, body=True, required_from=name)
+            if isinstance(named_cond, spack.spec.Spec):
+                requirements = self.spec_clauses(named_cond, body=True, required_from=name)
+            else:
+                requirements = named_cond
```

Comment:
I still think this is confusing, and mixing abstraction levels. We should find a way to split it in a neater way, without if/else here and with the assumption that client code is responsible to call `spec_clauses`.

Reply:
One idea is to change:
```python
def condition(self, required_spec, imposed_spec=None, name=None, msg=None, node=False):
```
into:
```python
def condition(self, required_spec, imposed_spec=None, name=None, msg=None, transform_requirements=None, transform_imposed=None):
```
where we _inject_ the transformation functions into condition, if we need to add anything. I'll try that out and report in a comment.